### PR TITLE
feat(terraform): create cross-variable edges between resources from the same module

### PR DESCRIPTION
**Pull request by YaaraVerner**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Implement part of [issue-3886](https://github.com/bridgecrewio/checkov/issues/3886).
Create cross-variable edges for terraform resources that are on the same modules and connected via variables, locals and etc.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/graph_builder/local_graph.py [hunk 1]
```diff
@@ -172,90 +177,115 @@ def _build_edges(self) -> None:
...
+            referenced_vertices = get_referenced_vertices_in_value(
+                value=attribute_value,
+                aliases=aliases,
+                resources_types=resources_types,
+            )
+            for vertex_reference in referenced_vertices:
+                # for certain blocks such as data and resource, the block name is composed from several parts.
+                # the purpose of the loop is to avoid not finding the node if the name has several parts
+                sub_values = [remove_index_pattern_from_str(sub_value) for sub_value in vertex_reference.sub_parts]
+                for i, _ in enumerate(sub_values):
```
Comment (tronxd) [inline checkov/terraform/graph_builder/local_graph.py:197]:
```suggestion
                for i in range(sub_values):
```

File: checkov/terraform/graph_builder/local_graph.py [hunk 2]
```diff
@@ -60,6 +62,9 @@ def build_graph(self, render_variables: bool) -> None:
             renderer = TerraformVariableRenderer(self)
             renderer.render_variables_from_local_graph()
             self.update_vertices_breadcrumbs_and_module_connections()
+            if strtobool(os.getenv("CHECKOV_EXPERIMENTAL_CROSS_VARIABLE_EDGES", "False")):
```
Comment (gruebel) [inline checkov/terraform/graph_builder/local_graph.py:65]: nice 🥇

File: checkov/terraform/graph_builder/local_graph.py [hunk 3]
```diff
@@ -172,90 +177,115 @@ def _build_edges(self) -> None:
...
+                            try:
+                                self._connect_module(
+                                    sub_values, attribute_key, self.vertices[dest_node_index], origin_node_index,
+                                    cross_variable_edges
                                 )
+                            except Exception as e:
+                                logging.warning(
+                                    f"Module {self.vertices[dest_node_index]} does not have source attribute, skipping"
+                                )
+                                logging.warning(e, stack_info=True)
```
Comment (gruebel) [inline checkov/terraform/graph_builder/local_graph.py:225]:
```suggestion
                                logging.warning(
                                    f"Module {self.vertices[dest_node_index]} does not have source attribute, skipping", exc_info=True
                                )
```
I think this should be the same, no need to have two logging calls

File: checkov/terraform/graph_builder/local_graph.py [hunk 4]
```diff
@@ -172,90 +177,115 @@ def _build_edges(self) -> None:
...
+            self._build_edges_for_vertex(origin_node_index, vertex, aliases, resources_types, True)
+
+    def _create_edge(self, origin_vertex_index: int, dest_vertex_index: int, label: str,
+                     cross_variable_edges: bool = False) -> bool:
         if origin_vertex_index == dest_vertex_index:
-            return
+            return False
         edge = Edge(origin_vertex_index, dest_vertex_index, label)
+        if cross_variable_edges:
+            if any(str(e) == str(edge) for e in self.edges):
```
Comment (gruebel) [inline checkov/terraform/graph_builder/local_graph.py:279]:
```suggestion
            if any(e == edge for e in self.edges):
```
the `__eq__` in class `Edge` does it already, so no need to do it here 🙂 
```suggestion
            if edge in self.edges:
```
I think this should even work

**Review by gruebel — COMMENTED** [review]
great work, added 2 suggestions

**Review by gruebel — APPROVED** [review]
🍰

File: checkov/terraform/graph_builder/local_graph.py [hunk 5]
```diff
@@ -172,90 +177,114 @@ def _build_edges(self) -> None:
...
+            referenced_vertices = get_referenced_vertices_in_value(
+                value=attribute_value,
+                aliases=aliases,
+                resources_types=resources_types,
+            )
+            for vertex_reference in referenced_vertices:
+                # for certain blocks such as data and resource, the block name is composed from several parts.
+                # the purpose of the loop is to avoid not finding the node if the name has several parts
+                sub_values = [remove_index_pattern_from_str(sub_value) for sub_value in vertex_reference.sub_parts]
+                for i in range(sub_values):
```
Comment (gruebel) [inline checkov/terraform/graph_builder/local_graph.py:197]:
```suggestion
                for i in range(len(sub_values)):
```
that fixes your integration issues

**Pull request merged**